javascript/lang/security/audit/prototype-pollution/prototype-pollution-function.yaml

```diff
@@ -68,6 +68,24 @@ rules:
                     $NAME.call(...)
                     ...
                   }
+              - pattern-inside: |
+                  $OBJ = function $NAME(...) {
+                    ...
+                    $NAME(...)
+                    ...
+                  }
+              - pattern-inside: |
+                  $OBJ = function $NAME(...) {
+                    ...
+                    $THIS.$NAME(...)
```

Comment:
Can this be this.$NAME?

Reply:
> @dottif I looked seeing how this would work in taint for the express example https://semgrep.dev/s/b1Rd I noticed that from the current JS file, it seems that one of the rules which is meant to be safe, seems unsafe?
> 
> ```
>     items[req.query.name] = req.query.text;
> ```
> 
> This technically should be an issue? Unless I am missing something
> 
> Overall your rule changes look good! I made one change to reduce a false positive, and left one comment if you could address it

The example of the `ok1` function is not vulnerable.
The user can't control the property name (it's hardcoded to "id"), as in the first example, and setting it to `__proto__`, so the variable `items` can't then refer to `Object.prototype`.

The `app.post` case is vulnerable, the check should be done on `id` not on `req.query.name`, also in `let items = req.session.todos[id1];` the `1` is a typo?

Also, by rewriting the rule that way (targeting only express cases) then it will not consider other assignments cases like
```javascript
\\two step
function foo(key1, key2, value) {
  var target = {};
  var tmp = target[key1];
  tmp[key2] = value;
}
```
Also, right now the assignment rule does not consider “direct” assignments that could lead to prototype pollution, and can only detect “two-step" assignments. I don't know if there is a way to match also the direct assignments without causing too many false positives.
```javascript
\\direct assignments 
function foo1(obj, a, b, val) {
  obj[a][b] = val;
}
function foo2(obj, a, b, c, val) {
  obj[a][b][c] = val;
}
```